source/agora/consensus/EnrollmentManager.d

```diff
@@ -568,6 +569,120 @@ public class EnrollmentManager
+
+        Returns:
+            true if the pre-image exists
+
+    ***************************************************************************/
+
+    public bool hasPreimage (const ref Hash enroll_key, ulong height) @trusted
+    {
+        auto results = this.db.execute("SELECT preimage from validator_set" ~
+                    " WHERE key = ?", enroll_key.toString());
+        if (!results.empty && results.oneValue!(byte[]).length != 0)
+        {
+            PreimageInfo loaded_image =
+                results.oneValue!(ubyte[]).deserializeFull!(PreimageInfo);
+            if (height <= loaded_image.height)
+                return true;
+        }
+        return false;
+    }
+
+    /***************************************************************************
+
+        Add a pre-image information to a validator data
+
+        Params:
+            preimage = the pre-image information to add
+
+        Returns:
+            true if the pre-image information has been added to the validator
+
+    ***************************************************************************/
+
+    public bool addPreimage (const ref PreimageInfo preimage) @safe nothrow
+    {
+        static ubyte[] buffer;
+
+        // check if the enrollment data exists
+        Enrollment stored_enroll;
+        try
+        {
+            if (!this.getEnrollment(preimage.enroll_key, stored_enroll))
+            {
+                log.info("Rejected adding a pre-image for non-existing" ~
+                    "enrollment, preimage:{}", preimage);
+                return false;
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation errort {}", ex);
+            return false;
+        }
+
+        // check if already exists
+        try
+        {
+            if (this.hasPreimage(preimage.enroll_key, preimage.height))
+            {
+                log.info("Rejected already existing pre-image, preimage:{}",
+                    preimage);
+                return false;
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation errort {}", ex);
+            return false;
+        }
+
+        // insert the pre-image into the table
+        buffer.length = 0;
+        () @trusted { assumeSafeAppend(buffer); } ();
+
+        scope SerializeDg dg = (scope const(ubyte[]) data) nothrow @safe
+        {
+            buffer ~= data;
+        };
+
+        try
+        {
+            serializePart(preimage, dg);
+        }
+        catch (Exception ex)
+        {
+            log.error("Serializtion error, addPreimage:{}, exception:{}",
+                preimage, ex);
+            return false;
+        }
+
+        try
+        {
+            () @trusted {
+                this.db.execute("UPDATE validator_set SET preimage = ? " ~
+                    "WHERE key = ?", buffer, preimage.enroll_key.toString());
```

Comment:
Same comment about memory allocation, `toString()` allocates. Can use `sformat` to write the entire string to a buffer before calling `execute`.